content/blog_posts/articles/introducing-tuyau.md

```diff
@@ -0,0 +1,301 @@
+
+Tuyau offers an automatically generated front-end client from your AdonisJS API that is 100% typesafe without the need to maintain any types or runtime code yourself. Tuyau uses code generation to detect input and output types for your routes. Let’s take the same API example, now using Tuyau:
+
+```ts
+// In your front-end
+import { createTuyau } from '@tuyau/client'
+import { api } from '@your-monorepo/my-adonisjs-app/.adonisjs/api'
+
+export const tuyau = createTuyau({
+  api,
+  baseUrl: 'http://localhost:3333',
+})
+
+const posts = tuyau.posts.$get({ page: 1, limit: 10 })
+const post = tuyau.posts({ id: 1 }).$get()
+```
+
+**Everything in this example is fully typesafe**: Route parameters (like `/posts/:id`), payloads, query params, and responses and errors. TypeScript prevents many errors: missing payload properties, forgotten request parameters, or typos in response data. TypeScript will scream at you, and your code won’t compile until you fix the errors.
+
+This also means you can use `tsc` in your CI pipeline to verify that your front end remains in sync with your backend. In short, Tuyau saves you time and reduces errors.
+
+If you’re familiar with [tRPC](https://trpc.io/), [Elysia Eden](https://elysiajs.com/eden/overview), or [Hono RPC](https://hono.dev/docs/guides/rpc), it’s the same concept.
+
+The Tuyau client works with any stack, including React, React Native, Vue, and Node.js. However, for it to function, you will need a monorepo setup.
+
+## Installation
+
+To get started, install Tuyau in your AdonisJS API:
+
+```bash
+node ace add @tuyau/core
+```
+
+Then, install the client in your frontend:
+
+```bash
+npm install @tuyau/client
+```
+
+Now, you’ll need to generate the types. The `@tuyau/core` package exposes a command, `node ace tuyau:generate`, which will create a `.adonisjs/api.ts` file at the root of your AdonisJS project. This file contains all the types necessary for the client to function. This command does **not** run automatically at the moment. You will need to run it manually after certain changes in your AdonisJS project, such as:
+
+- Adding a new route/controller to your project
+- Adding a `request.validateUsing` call in your controller method
+
+Other than that, you won’t need to run this command frequently. For example, if you update the controller method's return type or the Vine schema, you **don’t need** to run the command again.
+
+Once done, you can import and configure the client in your front-end as follows:
+
+```ts
+/// <reference path="../../adonisrc.ts" />
+
+import { createTuyau } from '@tuyau/client'
+import { api } from '@your-monorepo/server/.adonisjs/api'
+
+export const tuyau = createTuyau({
+  api,
+  baseUrl: 'http://localhost:3333',
+})
+```
+
+You’ll need to configure your mono repo so that your front end depends on your backend when importing the `.adonisjs/api.ts` file. More information [here](https://tuyau.julr.dev/docs/installation#sharing-the-api-definition).
```

Comment:
```suggestion
You’ll need to configure your monorepo so that your front end depends on your backend when importing the `.adonisjs/api.ts` file. More information [here](https://tuyau.julr.dev/docs/installation#sharing-the-api-definition).
```